### PR TITLE
Webex Feature/API

### PR DESCRIPTION
**Pull request by eumel8**
Signed-off-by: Frank Kloeker <f.kloeker@telekom.de>

## Description

Prometheus Alertmanager brings the functionality of Webex Notification. Start of Operator implementation was [here](https://github.com/prometheus-operator/prometheus-operator/pull/5254)
This is the part of API/CRD extension

## Type of change

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

ref:https://github.com/prometheus-operator/prometheus-operator/issues/5295

## Changelog entry

```release-note
Add Webex configs to AlertmanagerConfig CRD
```

### CODE REVIEW

**Comment by eumel8** [comment]
the bundle version needs to bump up

**Comment by xiu** [comment]
Great!

@eumel8 you can generate the bundle by using `make generate`. To help with formatting as well, `make format` and to check the formatting of the docs, `make check-docs` (https://github.com/prometheus-operator/prometheus-operator#contributing).

I see your tests are not passing, `make test-unit` can help there (https://github.com/prometheus-operator/prometheus-operator#testing).

**Comment by simonpasquier** [comment]
Thanks for starting this! You'd need to update https://github.com/prometheus-operator/prometheus-operator/blob/main/pkg/apis/monitoring/v1beta1/conversion_from.go and https://github.com/prometheus-operator/prometheus-operator/blob/main/pkg/apis/monitoring/v1beta1/conversion_to.go to convert between v1alpha1 and v1beta1 types.

File: pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go:581]: room_id is a mandatory field.

```suggestion
	// ID of the Webex Teams room where to send the messages.
	// +kubebuilder:validation:MinLength=1
	RoomID       string      `json:"roomID,omitempty"`
```

File: pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go:579]:
```suggestion
	// Whether to notify about resolved alerts.
	SendResolved *bool       `json:"sendResolved,omitempty"`
```

File: pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go:580]:
```suggestion
	// The Webex Teams API URL if different from the default API URL.
	APIURL       string      `json:"apiURL,omitempty"`
```

File: pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go:583]:
```suggestion
	// HTTP client configuration.
	HTTPConfig   *HTTPConfig `json:"httpConfig,omitempty"`
```

**Comment by eumel8** [comment]
thx for starting review, @xiu @simonpasquier

File: pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go:582]:
```suggestion
	// +kubebuilder:validation:MinLength=1
	RoomID       string      `json:"roomID"`
```

File: pkg/apis/monitoring/v1beta1/alertmanager_config_types.go
Comment (simonpasquier) [inline pkg/apis/monitoring/v1beta1/alertmanager_config_types.go:575]:
```suggestion
	// +kubebuilder:validation:MinLength=1
	RoomID       string      `json:"roomID"`
```

**Review by simonpasquier — CHANGES_REQUESTED** [review]
Can you add comments for each field? Next step is to wire the webex integration in https://github.com/prometheus-operator/prometheus-operator/blob/6be89b4010d06cbd2561b6650b6399438900a944/pkg/alertmanager/amcfg.go#L464

**Comment by eumel8** [comment]
just wondering, `make generate` fails on Ubuntu:

```
mv github.com/prometheus-operator/prometheus-operator/pkg/client/{versioned,informers,listers} pkg/client
mv: cannot stat 'github.com/prometheus-operator/prometheus-operator/pkg/client/informers': No such file or directory
mv: cannot stat 'github.com/prometheus-operator/prometheus-operator/pkg/client/listers': No such file or directory
make: *** [Makefile:142: k8s-client-gen] Error 1
```

**Comment by xiu** [comment]
> just wondering, `make generate` fails on Ubuntu:
> 
> ```
> mv github.com/prometheus-operator/prometheus-operator/pkg/client/{versioned,informers,listers} pkg/client
> mv: cannot stat 'github.com/prometheus-operator/prometheus-operator/pkg/client/informers': No such file or directory
> mv: cannot stat 'github.com/prometheus-operator/prometheus-operator/pkg/client/listers': No such file or directory
> make: *** [Makefile:142: k8s-client-gen] Error 1
> ```

This is not the issue as I see it. Check https://github.com/prometheus-operator/prometheus-operator/actions/runs/4088515306/jobs/7050273000, the step is checking whether all generated content has been committed. It thus runs `make generate` and does a global diff. In this case, it found changes.

Can you run `make generate` and commit the changes to your branch and try again?

**Comment by eumel8** [comment]
@xiu This was the error on my local Ubuntu WSL. I couldn't found out the cause and made make generate on another machine. Maybe there are some hints for build requirements beside make and Go.

**Comment by xiu** [comment]
> @xiu This was the error on my local Ubuntu WSL. I couldn't found out the cause and made make generate on another machine. Maybe there are some hints for build requirements beside make and Go.

Oh, I don't have a Windows machine to try that out unfortunately. Glad you made it work!

**Comment by eumel8** [comment]
@simonpasquier I'm not sure, anything else to change here?

**Comment by simonpasquier** [comment]
> @simonpasquier I'm not sure, anything else to change here?

You need to update the [convertReceiver()](https://github.com/prometheus-operator/prometheus-operator/blob/6be89b4010d06cbd2561b6650b6399438900a944/pkg/alertmanager/amcfg.go#L464) method to convert the Webex configuration from the CRD to the native Alertmanager configuration.

You can take a look at https://github.com/prometheus-operator/prometheus-operator/pull/4726 which did the same thing but for the Telegram integration. Does it make sense?

**Comment by fctb** [comment]
any time when it will get unblocked?

**Comment by eumel8** [comment]
@fctb It's progressing slowly. I'm a bit overwhelmed with the feature. Any help is appreciated, hopefully it will be over soon.

**Comment by puffitos** [comment]
Hi @simonpasquier, I'll be picking up from where @eumel8 left off. Please let me know if there's still something to be done. I've gone through the telegram MR and it seems to me we've covered everything. The changes you'd requested are also already implemented. I've just pulled the latest commits from the upstream and fixed a logic bug in the validation. And also a small grammar mistake I found :)

The only thing I've noticed is that the E2E tests don't successfully run on my ubuntu pc. The webex ones do, but some fail, unfortunately.

File: pkg/alertmanager/validation/v1alpha1/validation.go
Comment (ArthurSens) [inline pkg/alertmanager/validation/v1alpha1/validation.go:317]: The API definition already says that RoomID is a required field, right? Do we need this validation anyway?

**Review by ArthurSens — COMMENTED** [review]
Thanks @eumel8 for starting and @puffitos for picking it up, I think it's almost done :)

File: pkg/apis/monitoring/v1beta1/alertmanager_config_types.go
Comment (puffitos) [reply pkg/apis/monitoring/v1beta1/alertmanager_config_types.go]: I've changed the optional fields to pointers and accommodated the `convertWebexConfig` function to work with the new definition.

File: pkg/alertmanager/validation/v1alpha1/validation.go
Comment (puffitos) [reply pkg/alertmanager/validation/v1alpha1/validation.go]: Good point; i've removed the unnecessary check :+1:

File: pkg/apis/monitoring/v1beta1/alertmanager_config_types.go
Comment (puffitos) [reply pkg/apis/monitoring/v1beta1/alertmanager_config_types.go]: I've also changed the internal `webexConfig` struct in the `alertmanager` package to use pointers. 

E2E tests are still failing. The YAML config can't be parsed for the following reason:

```
ts=2023-06-17T06:10:30.122Z caller=coordinator.go:118 level=error component=configuration msg="Loading configuration file failed" file=/etc/alertmanager/config_out/alertmanager.env.yaml err="missing webex_configs.http_config.authorization"
```

Why is this happening? The e2e tests for webex were written in a similar fashion to the telegram ones, and the e2e were passing before moving to pointers.

Comment (puffitos) [reply pkg/apis/monitoring/v1beta1/alertmanager_config_types.go]: OK... I needed to take a look on how the global config for the tests is being created, then everything was clear. I had to make some changes:

I had to change the HTTPConfig field from optional to mandatory, as the Bearer Token of the Bot is expected to be there. The AlertManager code always checks for its presence:

https://github.com/prometheus/alertmanager/blob/90a8acd27661db09c9d9f1b4a2d4574a24466c76/config/notifiers.go#L196-L213
```go
// UnmarshalYAML implements the yaml.Unmarshaler interface.
func (c *WebexConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
	*c = DefaultWebexConfig
	type plain WebexConfig
	if err := unmarshal((*plain)(c)); err != nil {
		return err
	}

	if c.RoomID == "" {
		return fmt.Errorf("missing room_id on webex_config")
	}

	if c.HTTPConfig == nil || c.HTTPConfig.Authorization == nil {
		return fmt.Errorf("missing webex_configs.http_config.authorization")
	}

	return nil
}
``` 

I decided against expecting from the user to put the Authorization for the webex bot into their global config and made the http_config a mandatory field.

After that, the tests were still failing, because of missing smpt fields in the global AlertManager config. I've added an smpt_smarthost and a smpt_from, which are also mandatory fields for a global config (the config couldn't be unmarshalled for else wise).

Finally, I noticed that the `testPromOperatorStartsWithoutScrapeConfigCRD` was still failing. The test was checking whether the returned map with pod restarts of the prometheus operator pod was empty:

```goland
	require.Emptyf(t, restarts, "expected to get 1 container but got %d", len(restarts))
// this produced the funny looking error: expected to get 1 container but got 1
```

I've changed this to check whether only 1 container was really returned.

Sorry the long comment; just wanted to make sure I've explained all the changes.

**Comment by ArthurSens** [comment]
Nice work @puffitos! I want to apologize, I'm a bit busy this week and I'm struggling to find time to give a proper review

You found a tricky situation there with the HTTP Config! As you noticed, HTTP Config can be optional in WebexConfig if we follow Alertmanager's API definition because it can assume the value of Global HTTP Config. 

You give an interesting proposal here by making it required, it would certainly avoid that we have a nil HTTP config in case both Webex and Global HTTP config are nil, but we're breaking the case where the user really wants to re-use the HTTP config from Global config.

I think we'll need to keep HTTP Config as an optional field but validate it when reconciling the CRD.

Any other ideas @prometheus-operator/prometheus-operator-reviewers ?

File: pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go:581]: Let's create a URL type for proper validation
```suggestion
	APIURL *URL `json:"apiURL,omitempty"`
```

then

````
// +kubebuilder:validation:Pattern=`^https?://.+$`
type URL string
````

File: pkg/alertmanager/types.go
Comment (simonpasquier) [inline pkg/alertmanager/types.go:359]: I wouldn't change the types here as they come from the Alertmanager upstream configuration format.

File: pkg/alertmanager/operator.go
Comment (simonpasquier) [inline pkg/alertmanager/operator.go:1413]: we need a check on the amVersion too (e.g. return an error if version < v0.25.0 similar to the check in `checkTelegramConfigs()`).

File: pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go
Comment (puffitos) [reply pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go:581]: Should I also replace other URL-kind fields as with the new URL type (maybe in a separate PR)?

File: test/e2e/scrapeconfig_test.go
Comment (ArthurSens) [inline test/e2e/scrapeconfig_test.go:219]: Actually we're counting the amount of restarts here, not the amount of containers. The error message is certainly misleading!

That said, the test is correct. We don't expect restarts here

File: pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go
Comment (simonpasquier) [reply pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go:581]: > Does a dedicated URL type offer any other advantages, which I'm not aware of?

yes, `kubebuilder:validation:Pattern` is applied by the Kubernetes API without requiring any validating webhook configuration.

**Comment by simonpasquier** [comment]
@eumel8 I've added a few commits on top of your PR to address my comments.

File: test/e2e/scrapeconfig_test.go
Comment (puffitos) [inline test/e2e/scrapeconfig_test.go]: This test was failing on my machine with the message "expected to get 1 container but got 1". On line 213 one Pod is explicitly expected.  The GetPodRestartCount will therefore always return a map with at least one container, which should have 0 restarts. The `restarts` map has therefore a length of at least 1. 

Emptyf checks whether the map is empty, which it shouldn't. The error msg is even more cryptic, as it refers to the length of the map, which should be one according to the message. But here we're checking whether the map's nil / empty? 

I can't get my head around as to how this tests works :smile: I just wanted to explain, why I made this change and I why I didn't consider it unnecessary.

File: test/framework/resources/alertmanager-main-secret.yaml
Comment (puffitos) [reply test/framework/resources/alertmanager-main-secret.yaml:1]: I've reverted the change - if the E2E suite passes, it is indeed not necessary. I'd only changed it, as it didn't work when I ran the E2E tests locally and after debugging I'd found out the issue were the missing fields in the AM Config :shrug:

Comment (puffitos) [reply test/framework/resources/alertmanager-main-secret.yaml:1]: Tests passes so thanks for reviewing this change as well.

**Review by ArthurSens — APPROVED** [review]

File: pkg/alertmanager/operator.go
Comment (simonpasquier) [inline pkg/alertmanager/operator.go:1454]: (nit)
```suggestion
		return fmt.Errorf(`webexConfigs' is available in Alertmanager >= 0.25.0 only - current %s`, amVersion)
```

**Review by simonpasquier — COMMENTED** [review]
very good! One last nit...

**Review by ArthurSens — APPROVED** [review]
Oof, I approved and forgot to merge this one after 3 weeks... sorry! 😅 

Approving again and I'll let simon take another look if he wants to :)

I'm putting a reminder to merge it this week if there is no more change requests

**Pull request merged**
